hydra-cluster/bench/Bench/EndToEnd.hs

```diff
@@ -159,15 +169,18 @@ bench timeoutSeconds workDir dataset@Dataset{clientDatasets} clusterSize =
 -- @@
 --
 -- TODO: add more data points for memory and network consumption
-withOSStats :: FilePath -> IO () -> IO ()
+withOSStats :: FilePath -> IO a -> IO a
 withOSStats workDir action =
   findExecutable "dstat" >>= \case
     Nothing -> action
     Just exePath ->
       withCreateProcess (process exePath){std_out = CreatePipe} $ \_stdin out _stderr _processHandle ->
-        race_
+        race
           (collectStats out $ workDir </> "system.csv")
           action
+          >>= \case
+            Left () -> failure "stats collection process failed unexpectedly"
```

Comment:
I am not sure I understand what you are suggesting here. `collectStats` indeed "should" never return but by throwing an exception which won't reach that point. However, I need the result of the action and `race` returns an `Either a b` hence why the `failure`. I could use `absurd` as we do in the `HydraNode` module if you feel this is better.

Reply:
Unfortunately, using `Void` does not work because `forever` has type `IO () -> IO ()` so the first branch of `race` can indeed return `()`. Apart from a plain `error` I don't see much of another solution than failing here.